### PR TITLE
[training] fix: Preserve packed SFT metadata on middle PP stages

### LINKED ISSUE #2952: [bug] Abnormal SFT results with Megatron Bridge and packed datasets.
### Problem

I am fine-tuning a DeepSeek-V3 structure MoE model using SFT， it has 60 layers， include 1 dense layer and 59 moe layer. The training works correctly in the nemo2509 image (NeMo framework), but I encountered abnormal loss after migrating to the nemo2602 image (Megatron Bridge framework).
Below are the initial loss comparisons under different data processing scenarios (using a single data sample repeated 2,000 times):
Test 1: Single data sample repeated 2,000 times, then packed by the framework
nemo2509: 0.5716
nemo2602: 0.0347 (Significant anomaly)
Test 2: Single data sample repeated 2,000 times, non-packed
nemo2509: 0.5714
nemo2602: 0.5724
Observation: Loss is correctly aligned
Test 3: Single data sample packed first, then repeated 2,000 times
nemo2509: 0.5724
nemo2602: 0.5711
Observation: Loss is roughly aligned.
Conclusion:
The loss values are consistent in non-packed scenarios, but a major discrepancy occurs in Test 1. This indicates that the issue is likely related to the sequence packing logic within the Megatron Bridge in nemo2602.

I saved and compared the output tensors of the Attention and MLP blocks for each layer across both frameworks. My findings are as follows:
At the first layer(dense layer), the output is identical  in both frameworks.
At the second layer, the output of the Attention block is identical in both frameworks.
However, the output of the MLP block (MoE layer) differs starting from the second layer.
Upon further investigation, I discovered that the probs and routing_map returned by the MoELayer::route() function in the second layer are inconsistent between the two frameworks, which leads to the divergent MLP outputs.

### Minimal repro

```shell
this is my transformer config:

tensor_model_parallel_size: 8
pipeline_model_parallel_comm_backend: None
pipeline_model_parallel_size: 12
virtual_pipeline_model_parallel_size: None
sequence_parallel: True
context_parallel_size: 1
hierarchical_context_parallel_sizes: None
max_seqlen_per_dp_cp_rank: None
hybrid_context_parallel: False
expert_model_parallel_size: 8
expert_tensor_parallel_size: 1
moe_extended_tp: False
perform_initialization: True
use_cpu_initialization: False
fp16: False
bf16: True
params_dtype: torch.float32
timers: None
finalize_model_grads_func: None
grad_scale_func: None
no_sync_func: None
grad_sync_func: None
param_sync_func: None
deterministic_mode: False
enable_autocast: False
autocast_dtype: torch.bfloat16
num_microbatches_with_partial_activation_checkpoints: None
gradient_accumulation_fusion: False
async_tensor_model_parallel_allreduce: True
use_te_rng_tracker: False
tp_comm_overlap: False
tp_comm_bulk_wgrad: True
tp_comm_bulk_dgrad: True
tp_comm_overlap_ag: True
tp_comm_overlap_rs: True
tp_comm_overlap_rs_dgrad: False
tp_comm_split_ag: True
tp_comm_atomic_ag: False
tp_comm_split_rs: True
tp_comm_atomic_rs: False
cross_entropy_loss_fusion: False
cross_entropy_fusion_impl: te
tp_comm_overlap_disable_qkv: False
tp_comm_overlap_disable_fc1: False
tp_comm_bootstrap_backend: nccl
overlap_moe_expert_parallel_comm: False
delay_wgrad_compute: False
ep_overlap_early_attn_memory_release: False
pipeline_dtype: torch.bfloat16
variable_seq_lengths: False
overlap_p2p_comm: False
batch_p2p_comm: True
batch_p2p_sync: True
use_ring_exchange_p2p: False
deallocate_pipeline_outputs: False
defer_embedding_wgrad_compute: False
wgrad_deferral_limit: -1
overlap_p2p_comm_warmup_flush: False
microbatch_group_size_per_vp_stage: 12
mtp_standalone: False
cpu_offloading: False
cpu_offloading_num_layers: 0
_cpu_offloading_context: None
cpu_offloading_activations: True
cpu_offloading_weights: False
cpu_offloading_double_buffering: False
barrier_with_L1_time: True
num_layers: 60
mtp_num_layers: None
mtp_loss_scaling_factor: 0.1
num_layers_in_first_pipeline_stage: None
num_layers_in_last_pipeline_stage: None
pipeline_model_parallel_layout: None
account_for_embedding_in_pipeline_split: False
account_for_loss_in_pipeline_split: False

attention_backend: AttnBackend.auto
softmax_scale: None
softmax_type: vanilla
hidden_dropout: 0.0
attention_dropout: 0.0
fp32_residual_connection: False
apply_residual_connection_post_layernorm: False
layernorm_epsilon: 1e-06
layernorm_zero_centered_gamma: False
add_bias_linear: False
add_qkv_bias: False
gated_linear_unit: True
activation_func: <function silu at 0x7fb3b6affc40>
activation_func_fp8_input_store: False
glu_linear_offset: 0.0
activation_func_clamp_value: None
rotary_interleaved: False
window_size: None
window_attn_skip_freq: None
normalization: RMSNorm
qk_layernorm: True
qk_l2_norm: False
qk_clip: False
qk_clip_alpha: 0.5
qk_clip_threshold: 100
log_max_attention_logit: False
attention_output_gate: False
test_mode: False
calculate_per_token_loss: False
multi_latent_attention: True
no_rope_freq: None
experimental_attention_variant: None
dsa_indexer_n_heads: None
dsa_indexer_head_dim: None
dsa_indexer_topk: None
dsa_indexer_loss_coeff: None
dsa_indexer_use_sparse_loss: False
linear_attention_freq: None
linear_conv_kernel_dim: 4
linear_key_head_dim: 128
linear_value_head_dim: 128
linear_num_key_heads: 16
linear_num_value_heads: 32
init_method: functools.partial(<function normal_ at 0x7fb3b6ba05e0>, mean=0.0, std=0.006)
output_layer_init_method: functools.partial(<function normal_ at 0x7fb3b6ba05e0>, mean=0.0, std=0.0005477225575051661)
init_method_std: 0.006
embedding_init_method: functools.partial(<function normal_ at 0x7fb3b6ba05e0>, mean=0.0, std=0.006)
embedding_init_method_std: 0.006
init_model_with_meta_device: False
apply_query_key_layer_scaling: False
attention_softmax_in_fp32: True
disable_bf16_reduced_precision_matmul: False
bias_activation_fusion: False
masked_softmax_fusion: False
persist_layer_norm: False
memory_efficient_layer_norm: False
bias_dropout_fusion: False
apply_rope_fusion: False
use_fused_weighted_squared_relu: False
fused_single_qkv_rope: False
recompute_granularity: full
recompute_method: uniform
recompute_num_layers: 1
distribute_saved_activations: False
recompute_modules: ['core_attn']
fp8: None
fp8_recipe: tensorwise
fp8_param: False
fp8_quantizer_factory: None
fp8_margin: 0
fp8_interval: 1
fp8_amax_history_len: 1
fp8_amax_compute_algo: most_recent
fp8_wgrad: True
fp8_dot_product_attention: False
fp8_multi_head_attention: False
tp_only_amax_red: False
first_last_layers_bf16: False
num_layers_at_start_in_bf16: 0
num_layers_at_end_in_bf16: 0
use_kitchen: False
use_kitchen_attention: False
kitchen_attention_backend: sdpa
fp4: None
fp4_recipe: nvfp4
fp4_param: False
fp4_quantizer_factory: None
moe_shared_expert_intermediate_size: 3072
moe_shared_expert_gate: False
moe_shared_expert_overlap: False
moe_layer_freq: [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
moe_ffn_hidden_size: 1536
moe_router_load_balancing_type: seq_aux_loss
moe_router_topk: 6
moe_enable_routing_replay: False
moe_router_topk_limited_devices: None
moe_router_padding_for_quantization: False
moe_router_padding_for_fp8: False
moe_router_num_groups: 1
moe_router_group_topk: 1
moe_router_pre_softmax: True
moe_router_topk_scaling_factor: 2.827
moe_router_score_function: sigmoid
moe_router_dtype: fp32
moe_router_enable_expert_bias: True
moe_router_bias_update_rate: 0.001
moe_router_force_load_balancing: False
moe_grouped_gemm: True
moe_use_legacy_grouped_gemm: False
moe_aux_loss_coeff: 0.0001
moe_z_loss_coeff: None
moe_input_jitter_eps: None
moe_token_dropping: False
moe_token_dispatcher_type: alltoall
moe_enable_deepep: False
moe_flex_dispatcher_backend: deepep
moe_per_layer_logging: False
moe_expert_capacity_factor: None
moe_pad_expert_input_to_capacity: False
moe_token_drop_policy: probs
moe_layer_recompute: False
moe_permute_fusion: False
moe_router_fusion: False
moe_apply_probs_on_input: False
moe_latent_size: None
moe_deepep_num_sms: 20
moe_hybridep_num_sms: 16
cp_comm_type: None
enable_cuda_graph: False
cuda_graph_use_single_mempool: False
cuda_graph_retain_backward_graph: False
cuda_graph_warmup_steps: 3
external_cuda_graph: False
cuda_graph_impl: none
cuda_graph_scope: []
clone_scatter_output_in_embedding: True
disable_parameter_transpose_cache: False
config_logger_dir: 
flash_decode: False
batch_invariant_mode: False
use_te_activation_func: False
inference_rng_tracker: False
inference_sampling_seed: 42
symmetric_ar_type: None
use_inference_optimized_layers: False
inference_fuse_tp_communication: False
mrope_section: None
is_hybrid_model: False
mamba_state_dim: 128
mamba_head_dim: 64
mamba_num_groups: 8
mamba_num_heads: None
use_mamba_mem_eff_path: True
mlp_chunks_for_prefill: 1
heterogeneous_block_specs: False
hetereogenous_dist_checkpoint: False
quant_recipe: None
transformer_impl: transformer_engine
fine_grained_activation_offloading: False
offload_modules: []
min_offloaded_tensor_size: 1048576
rope_type: rope
rotary_base: 1000000
rotary_percent: 1.0
rotary_scaling_factor: 1.0
original_max_position_embeddings: 4096
beta_fast: 32
beta_slow: 1
mscale: 1.0
mscale_all_dim: 1.0
cache_mla_latents: False
fp16_lm_cross_entropy: False
parallel_output: True
share_embeddings_and_output_weights: False
make_vocab_size_divisible_by: 512
position_embedding_type: rope
seq_len_interpolation_factor: None
seq_length: 32768
scatter_embedding_sequence_parallel: True
tp_comm_overlap_cfg: None
use_transformer_engine_full_layer_spec: False
use_transformer_engine_op_fuser: False
transformer_layer_spec: functools.partial(<function get_gpt_decoder_block_spec at 0x7fb0dcd2d620>, use_transformer_engine=True)
hf_model_id: None
generation_config: None
vocab_size: 152064
should_pad_vocab: False
mtp_enabled: False
restore_modelopt_state: False
use_arbitrary_attention_mask: None
max_position_embeddings: 32768
```

### Expected behavior

The initial loss in nemo2602's Megatron Bridge aligns with nemo2509 when using packed data

### Affected area

area:training

### Regression?

Yes

### Environment

_No response_

### Logs

```shell

```

### PR DESCRIPTION
**Pull request by yaoyu-33**
## Summary

Closes #2952.

Fixes GPT packed SFT batch generation for pipeline-parallel middle stages. Middle PP stages now keep the existing all-None fast path for non-packed/default attention, but load attention metadata when packed sequence specs or dataset-provided attention masks are active. This preserves `cu_seqlens`, `cu_seqlens_argmin`, `max_seqlen`, optional unpadded cu-seqlens, and explicit attention masks through transformer middle stages.

Also avoids hybrid-model `total_tokens` crashes on metadata-only middle stages by deriving the value from packed cu-seqlens when token/label tensors are absent.

## CW evidence

Pre-fix deterministic repro failed on cw before patching:
- Job: `11976945`
- Log: `/lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_llm/users/yuya/logs/issue2952-mira/pre_repro_11976945.log`
- Command: `uv run --active --no-sync python -m pytest tests/unit_tests/training/test_gpt_step.py -v`
- Result: `test_middle_pp_stage_preserves_packed_metadata` failed because `out_cu_seqlens` was `None` on a middle PP stage.

Post-fix rerun of the same repro passed:
- Job: `11977013`
- Log: `/lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_llm/users/yuya/logs/issue2952-mira/post_repro_11977013.log`
- Result: `21 passed`

Additional packing validation on cw:
- Job: `11977108`
- Log: `/lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_llm/users/yuya/logs/issue2952-mira/post_packing_11977108.log`
- `uv run --active --no-sync python -m pytest tests/unit_tests/training/utils/test_packed_seq_utils.py -v`: `11 passed`
- `uv run --active --no-sync python -m pytest tests/unit_tests/training/test_config.py -k \"packed_sequence or pack_sequences_in_batch or context_parallel_seq_length_divisibility or context_parallel_finetuning_validations\" -v`: `14 passed, 159 deselected`

Full DeepSeek-V3 loss repro was not run; the targeted metadata repro directly exercises the suspected middle-PP metadata drop and avoids the much larger 12-way PP MoE SFT setup.

## Local checks

- `git diff --check`: passed
- `uv run --no-sync ruff check src/megatron/bridge/training/gpt_step.py tests/unit_tests/training/test_gpt_step.py`: passed
- `uv run --no-sync ruff format --check src/megatron/bridge/training/gpt_step.py tests/unit_tests/training/test_gpt_step.py`: passed
- `uv run pre-commit run --all-files`: blocked by local NRX wheel platform resolution (`manylinux_2_31` host vs `manylinux_2_39` wheel)
- `uv run --active --no-sync pre-commit run --all-files`: passed

### CODE REVIEW

File: src/megatron/bridge/training/gpt_step.py [hunk 1]
```diff
@@ -166,12 +184,13 @@ def get_batch(
...
     is_last = is_pp_last_stage(pg_collection.pp)
-    if (not is_first) and (not is_last):
+    is_middle = (not is_first) and (not is_last)
+    if is_middle and not _middle_pp_stage_needs_batch(cfg):
         return None, None, None, None, None, None, None, None, None, None
 
     batch = get_batch_from_iterator(
         data_iterator,
-        use_mtp,
+        use_mtp and not is_middle,
```
Comment (yaoyu-33) [inline src/megatron/bridge/training/gpt_step.py]: Is this correct? we are doing some hacks on top of use_mtp, not sure it's best way to handle, we can rename use_mtp in args though.

File: src/megatron/bridge/training/gpt_step.py [hunk 2]
```diff
@@ -258,7 +321,9 @@ def _forward_step_common(
         # which is only needed for Mamba/hybrid SSM layers. Skip it for pure
         # transformer models to avoid per-step CUDA overhead.
         if getattr(config, "is_hybrid_model", False):
-            packed_seq_params["total_tokens"] = tokens.size(1) if tokens is not None else labels.size(1)
+            packed_seq_params["total_tokens"] = _get_packed_total_tokens(
```
Comment (yaoyu-33) [inline src/megatron/bridge/training/gpt_step.py]: i tends to not do this for now, I think let's do if needs cuseqlens, let's just pass everything to middle layers as well for now

**Comment by yaoyu-33** [comment]
Follow-up loss-level ablation evidence for issue #2952 / PR #3926.

I ran a short internal validation run with `pipeline_model_parallel_size=4` on both the patched PR branch and an unpatched `origin/main` baseline. This used the issue #2952 attachments (`training.json`, `training_packed.tar.gz`, `sft.py`) as input artifacts; the actual training run was intentionally scaled down to a tiny random-init GPT config while preserving PP>2 and the packed metadata shape (`cu_seqlens` with the customer-style `-1` sentinel). Because this is not the full customer DeepSeek/SFT checkpoint setup, the absolute loss is random-init CE scale (~11.9), not the reported ~0.57 customer scale.

The internal validation run completed successfully.

| Case | Data construction | Branch | Packed? | First-step loss | Expected | Result |
| --- | --- | --- | --- | ---: | --- | --- |
| A | Single attached sample repeated, framework-packed in scaled harness | patched PR branch | yes | 11.926380 | Align with B/C | PASS; exactly matches C and is close to B |
| B | Single attached sample repeated, non-packed | patched PR branch | no | 11.909723 | Baseline aligned loss | PASS |
| C | Single attached sample packed first, then repeated | patched PR branch | yes | 11.926380 | Align with A/B | PASS; exactly matches A |
| A | Same framework-packed scaled construction | unpatched baseline | yes | 11.923078 | Comparison; full customer anomaly would be low only in exact setup | Aligned in scaled harness; did not reproduce 0.0347 anomaly |
| B | Same non-packed scaled construction | unpatched baseline | no | 11.909723 | Baseline aligned loss | PASS |
| C | Same packed-first scaled construction | unpatched baseline | yes | 11.926380 | Baseline aligned loss | PASS |

Notes:
- The patched PR branch shows the desired loss-level behavior in the PP=4 scaled repro: framework-packed and packed-first match exactly, and the non-packed baseline is in the same narrow band.
- The unpatched comparison was practical and completed, but this scaled random-init harness did not reproduce the customer's `nemo2602` low-loss anomaly. I would treat this as patched absolute-alignment evidence, not as an exact customer numerical reproduction.
- Earlier attempts were harness setup retries only and are not evidence runs.

File: src/megatron/bridge/training/gpt_step.py (hunk 1 above)
Comment (yaoyu-33) [reply src/megatron/bridge/training/gpt_step.py]: Updated in `2a85d5fbf153e9abbede6216c4324a6fef9de819`: `use_mtp` is passed through unchanged now. Middle-stage full-batch loading is controlled by an explicit `include_full_batch_fields` flag, so the MTP decision is no longer overloaded with the middle-PP packed path. Internal validation run passed the targeted training tests.

File: src/megatron/bridge/training/gpt_step.py (hunk 2 above)
Comment (yaoyu-33) [reply src/megatron/bridge/training/gpt_step.py]: Updated in `2a85d5fbf153e9abbede6216c4324a6fef9de819`: removed the metadata-only middle-PP path and the cu-seqlens-derived token-count helpers. Middle PP keeps the all-`None` fast path when no packed/custom attention metadata is needed; when packed/custom attention metadata is active, it now loads and forwards the full batch fields plus packed metadata. Internal validation run passed the targeted training tests.

**Comment by yaoyu-33** [comment]
customer verified fix.
